server/src/main/java/org/elasticsearch/cluster/ClusterState.java

```diff
@@ -158,8 +158,21 @@ public ClusterState(
         this.customs = customs;
         this.wasReadFromDiff = wasReadFromDiff;
         this.routingNodes = routingNodes;
-        assert routingNodes == null || routingNodes.equals(new RoutingNodes(this))
-            : "RoutingNodes [" + routingNodes + "] are not consistent with this cluster state [" + new RoutingNodes(this) + "]";
+        assert assertConsistentRoutingNodes(routingTable, nodes, routingNodes);
+    }
+
+    private static boolean assertConsistentRoutingNodes(
+        RoutingTable routingTable,
+        DiscoveryNodes nodes,
+        @Nullable RoutingNodes routingNodes
+    ) {
+        if (routingNodes == null) {
+            return true;
+        }
+        final RoutingNodes expected = new RoutingNodes(routingTable, nodes);
+        assert routingNodes.equals(expected)
```

Comment:
`equals` does not look like it compares all fields, for instance on `ShardRouting`. Also, it would be even nicer to verify that all the ShardRouting instances are the same?

 But then again, that really does not change with this PR, so feel free to disregard this.